File: languages/javascript-or-typescript/cross-language-features.js
const WillyHorizont = require("willyhorizont.github.io");
const { jsonStringify } = WillyHorizont.Utils;

/*
1. support closure as value, or has workaround
*/
const sayHello = (callbackFunction) => {
    console.log("hello");
    callbackFunction();
};
sayHello(() => {
    console.log("wold");
});
const createMultiplier = (aa) => (bb) => (aa * bb);
const multiplyByTwo = createMultiplier(2);
console.log(`multiply_by_two(10): ${multiplyByTwo(10)}`);
const multiplyByEight = createMultiplier(8);
console.log(`multiply_by_eight(4): ${multiplyByEight(4)}`);
console.log(`multiply_by_two(8): ${multiplyByTwo(8)}`);

/*
2. support dynamic-typed value, or has workaround
*/
const xlList = [
    null,
    true,
    false,
    "foo",
    0,
    -123,
    123.789,
    -123.789,
    [1, 2, 3],
    { "foo": "bar" },
    (aa, bb) => (aa * bb),
];
console.log(`xl_list: ${jsonStringify(xlList)}`);
console.log(`xl_list: ${jsonStringify(xlList, { pretty: true })}`);
const xlDict = {
    "xl_none": null,
    "xl_bool_true": true,
    "xl_bool_false": false,
    "xl_string": "foo",
    "xl_int_positive": 0,
    "xl_int_negative": -123,
    "xl_float_positive": 123.789,
    "xl_float_negative": -123.789,
    "xl_list": [1, 2, 3],
    "xl_dict": { "foo": "bar" },
    "xl_closure": (aa, bb) => (aa * bb),
};
console.log(`xl_dict: ${jsonStringify(xlDict)}`);
console.log(`xl_dict: ${jsonStringify(xlDict, { pretty: true })}`);

File: languages/matlab-or-octave/cross-language-features.m
addpath(fileparts(fullfile({fileparts(mfilename("fullpath"))}{1}, "..", "..", "runtimes", "matlab-or-octave", "willyhorizont", "runtime", "xl.m")));

%{
1. support lambda as value, or has workaround
%}
sayhello = @(callbackfunction) {
    disp("hello"),
    callbackfunction()
}{end};
sayhello(@() {
    disp("world")
}{end});
createmultiplier = @(aa) {@(bb) {aa * bb}{end}}{end};
multiplybytwo = createmultiplier(2);
disp(cstrcat("multiply_by_two(10): ", xl.jsonstringify(multiplybytwo(10))));
multiplybyeight = createmultiplier(8);
disp(cstrcat("multiply_by_eight(4): ", xl.jsonstringify(multiplybyeight(4))));
disp(cstrcat("multiply_by_two(8): ", xl.jsonstringify(multiplybytwo(8))));

%{
2. support dynamic-typed value, or has workaround
%}
xllist = {
    {},
    true,
    false,
    "foo",
    0,
    -123,
    123.789,
    -123.789,
    {1, 2, 3},
    xl.dict("foo", "bar"),
    @(aa, bb) {aa * bb}{end},
};
disp(cstrcat("xl_list: ", xl.jsonstringify(xllist)));
disp(cstrcat("xl_list: ", xl.jsonstringify(xllist, struct("pretty", true))));
xldict = xl.dict( ...
    "xl_none", {}, ...
    "xl_bool_true", true, ...
    "xl_bool_false", false, ...
    "xl_string", "foo", ...
    "xl_int_positive", 0, ...
    "xl_int_negative", -123, ...
    "xl_float_positive", 123.789, ...
    "xl_float_negative", -123.789, ...
    "xl_list", {1, 2, 3}, ...
    "xl_dict", xl.dict("foo", "bar"), ...
    "xl_lambda", @(aa, bb) {aa * bb}{end} ...
);
disp(cstrcat("xl_dict: ", xl.jsonstringify(xldict)));
disp(cstrcat("xl_dict: ", xl.jsonstringify(xldict, struct("pretty", true))));
